subsystems/provisioning/bluetooth/bluetooth_wifi_provisioner.go

```diff
@@ -0,0 +1,171 @@
+	} `json:"networks"`
+}
+
+func (awns *AvailableWiFiNetworks) ToBytes() ([]byte, error) {
+	return json.Marshal(awns)
+}
+
+// BluetoothWiFiProvisioner provides an interface for managing BLE (bluetooth-low-energy) peripheral advertisement on Linux.
+type BluetoothWiFiProvisioner struct{}
+
+// Start begins advertising a bluetooth service that acccepts WiFi and Viam cloud config credentials.
+func (bwp *BluetoothWiFiProvisioner) Start(ctx context.Context) error {
+	if err := bwp.startAdvertisingBLE(ctx); err != nil {
+		return err
+	}
+	bwp.enableAutoAcceptPairRequest() // Async goroutine (hence no error check) which auto-accepts pair requests on this device.
+	return nil
+}
+
+// Stop stops advertising a bluetooth service which (when enabled) accepts WiFi and Viam cloud config credentials.
+func (bwp *BluetoothWiFiProvisioner) Stop() error {
+	return bwp.stopAdvertisingBLE()
+}
+
+// Update updates the list of networks that are advertised via bluetooth as available.
+func (bwp *BluetoothWiFiProvisioner) RefreshAvailableNetworks(ctx context.Context, awns *AvailableWiFiNetworks) error {
+	return bwp.writeAvailableNetworks(ctx, awns)
+}
+
+// WaitForCredentials returns credentials, the minimum required information to provision a robot and/or its WiFi.
+func (bwp *BluetoothWiFiProvisioner) WaitForCredentials(
+	ctx context.Context, requiresCloudCredentials bool, requiresWiFiCredentials bool,
+) (*Credentials, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if !requiresWiFiCredentials && !requiresCloudCredentials {
+		return nil, errors.New("should be waiting for either cloud credentials or WiFi credentials")
+	}
+	var ssid, psk, robotPartKeyID, robotPartKey string
+	var ssidErr, pskErr, robotPartKeyIDErr, robotPartKeyErr error
+	wg := sync.WaitGroup{}
+	if requiresWiFiCredentials {
+		wg.Add(2)
+		utils.ManagedGo(
+			func() {
+				if ssid, ssidErr = retryCallbackOnExpectedError(
+					ctx, bwp.readSsid, &emptyBluetoothCharacteristicError{}, "failed to read ssid",
+				); ssidErr != nil {
+					cancel()
+				}
+			},
+			wg.Done,
+		)
+		utils.ManagedGo(
+			func() {
+				if psk, pskErr = retryCallbackOnExpectedError(
+					ctx, bwp.readPsk, &emptyBluetoothCharacteristicError{}, "failed to read psk",
+				); pskErr != nil {
+					cancel()
+				}
+
+			},
+			wg.Done,
+		)
+	}
+	if requiresCloudCredentials {
+		wg.Add(2)
+		utils.ManagedGo(
+			func() {
+				if robotPartKeyID, robotPartKeyIDErr = retryCallbackOnExpectedError(
+					ctx, bwp.readRobotPartKeyID, &emptyBluetoothCharacteristicError{}, "failed to read robot part key ID",
+				); robotPartKeyIDErr != nil {
+					cancel()
+				}
+			},
+			wg.Done,
+		)
+		utils.ManagedGo(
+			func() {
+				if robotPartKey, robotPartKeyErr = retryCallbackOnExpectedError(
+					ctx, bwp.readRobotPartKey, &emptyBluetoothCharacteristicError{}, "failed to read robot part key",
+				); robotPartKeyErr != nil {
+					cancel()
+				}
+			},
+			wg.Done,
+		)
+	}
```

Comment:
this pattern feels odd to me - why are we starting 4 goroutines that's each calling 1 function in a loop? Something useful for context would be to know on a high level how each of the functions will work

Reply:
I'm looking at https://github.com/viamrobotics/agent/pull/68/files so let me know if I'm missing something.

It seems possible to only have 1 goroutine that calls readSsid, readPsk, readRobotPartKeyID, and readRobotPartKey sequentially and then retry every second, is there a reason why not?